File: frontend/src/api/article.ts
import request from '@/utils/request'

export interface User {
  id: number
  username: string
  displayName?: string
  avatar?: string
}

export interface Category {
  id: number
  name: string
  description?: string
  icon?: string
}

export interface Tag {
  id: number
  name: string
  color?: string
}

export interface Article {
  id: number
  title: string
  content?: string
  summary?: string
  published?: boolean
  viewCount: number
  createdAt: string
  updatedAt?: string
  author?: User
  category?: Category
  tags?: Tag[]
}

export interface PageResult<T> {
  content: T[]
  totalElements: number
  totalPages: number
  number: number
  size: number
  first: boolean
  last: boolean
}

export const articleApi = {
  getPublishedArticles: (page = 0, size = 10, category?: number) => {
    let url = `/articles/public/list?page=${page}&size=${size}`
    if (category != null) url += `&category=${category}`
    return request.get(url)
  },
  getArticleById: (id: number | string) => request.get(`/articles/public/${id}`),
  getRecentArticles: () => request.get('/articles/public/recent'),
  getPopularArticles: () => request.get('/articles/public/popular'),
  createArticle: (data: Record<string, unknown>) => request.post('/articles', data),
  updateArticle: (id: number, data: Record<string, unknown>) => request.put(`/articles/${id}`, data),
  deleteArticle: (id: number) => request.delete(`/articles/${id}`),
}

export const categoryApi = {
  getAllCategories: () => request.get('/categories'),
  getCategoryById: (id: number) => request.get(`/categories/${id}`),
}

export const tagApi = {
  getAllTags: () => request.get('/tags'),
  getTagById: (id: number) => request.get(`/tags/${id}`),
}

File: api/app/schemas/article.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class CategoryRef(BaseModel):
    id: int | None = None
    name: str | None = None
    description: str | None = None
    icon: str | None = None
    sort_order: int | None = Field(None, alias="sortOrder")

    model_config = ConfigDict(populate_by_name=True)


class TagRef(BaseModel):
    id: int | None = None
    name: str | None = None
    color: str | None = None

    model_config = ConfigDict(populate_by_name=True)


class CategoryDTO(BaseModel):
    id: int
    name: str
    description: str | None = None
    icon: str | None = None
    sort_order: int = Field(0, serialization_alias="sortOrder")

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, by_alias=True)


class CategoryWrite(BaseModel):
    id: int | None = None
    name: str
    description: str | None = None
    icon: str | None = None
    sort_order: int = Field(0, alias="sortOrder")

    model_config = ConfigDict(populate_by_name=True)


class TagDTO(BaseModel):
    id: int
    name: str
    color: str | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, by_alias=True)


class TagWrite(BaseModel):
    id: int | None = None
    name: str
    color: str | None = None

    model_config = ConfigDict(populate_by_name=True)


class AuthorDTO(BaseModel):
    id: int
    username: str
    display_name: str | None = Field(None, serialization_alias="displayName")
    avatar: str | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, by_alias=True)


class ArticleDTO(BaseModel):
    id: int
    title: str
    content: str
    summary: str | None = None
    cover_image: str | None = Field(None, serialization_alias="coverImage")
    published: bool = False
    view_count: int = Field(0, serialization_alias="viewCount")
    created_at: datetime | None = Field(None, serialization_alias="createdAt")
    updated_at: datetime | None = Field(None, serialization_alias="updatedAt")
    category: CategoryDTO | None = None
    tags: list[TagDTO] | None = None
    author: AuthorDTO | None = None

    model_config = ConfigDict(from_attributes=True, populate_by_name=True, by_alias=True)


class ArticleWrite(BaseModel):
    id: int | None = None
    title: str
    content: str
    summary: str | None = None
    cover_image: str | None = Field(None, alias="coverImage")
    published: bool = False
    category: CategoryRef | None = None
    tags: list[TagRef] | None = None

    model_config = ConfigDict(populate_by_name=True)
